src/cpp/rtps/flowcontrol/FlowControllerImpl.hpp

```diff
@@ -0,0 +1,1489 @@
+#ifndef _RTPS_FLOWCONTROL_FLOWCONTROLLERIMPL_HPP_
+#define _RTPS_FLOWCONTROL_FLOWCONTROLLERIMPL_HPP_
+
+#include "FlowController.hpp"
+#include <fastdds/rtps/common/Guid.h>
+#include <fastdds/rtps/writer/RTPSWriter.h>
+
+#include <map>
+#include <unordered_map>
+#include <thread>
+#include <mutex>
+#include <cassert>
+#include <condition_variable>
+#include <chrono>
+
+namespace eprosima {
+namespace fastdds {
+namespace rtps {
+
+/** Auxiliary classes **/
+
+struct FlowQueue
```

Comment:
As this is handling several lists, similar code is being repeated. I would add an internal `struct List` with the corresponding `head` and `tail` fields, and methods to avoid code duplication.